### ISSUE TITLE
Extern Objects (aka Reference Passing)

### ISSUE THREAD

**Issue by dOrgJelli** [Web3-API/prototype#991]
I want to efficiently pass around a reference to a POD object stored externally. This can speed up passing the object from wasm module to wasm module, without having to serialize/deserialize each time.

Call these "extern objects". Extern objects are serialized as a unique IDs (u32?), and resolve "auto-magically" to data stored within the host (client).

For example, consider a sender and a receiver. Sender has large amount of data to pass through:
```graphql
type Data {
  lotsOfBytes: Bytes!
}

type Module {
  dispatchData(
    uris: [String!]!
    data: Data! @extern
  ): Data! @extern
}
```
```typescript
export function dispatchData(args) {
  // u32 unique storage/ref id
  args.data.__uuid

  // getter, loads reference data from host environment
  args.data.lotsOfBytes()

  for (const uri of args.uris) {
    // efficiently pass around a ref id for the data
    wrap_subinvoke(uri, "method", args.data);
  }
  
  // fully load (clone) the data from the host environment into wasm
  const data = Data.clone(args.data);

  // store a new reference within the client cache
  const ref = wrap_store(data.toBuffer());
}
```

omitted imports from above:
```typescript
import { Data } from "./wrap";
import {
  wrap_subinvoke,
  wrap_store
} from "@polywrap/wasm-runtime";
```
Need to work out what the wasm-runtime interface should look like :)

# Additional Feature
Access capabilities could be configured on a per-reference-pass basis. This can allow you to only give away access to specific sub-pieces of data. For example, I could give a module read capabilities of an object in storage, but no write. Additionally I can restrict on a per-property basis, or pass an "opaque reference" so that the receiver only has an ID and none of the underlying data.

# EDIT!
I'm realizing that capabilities should probably be spec'd out independent of this feature to see how it can best be implemented. Created an issue here: https://github.com/polywrap/technical-council/issues/61

**Comment by Niraj-Kamdar** [comment]
@dOrgJelli Should this be done pre-origin?

**Comment by dOrgJelli** [comment]
No this is too large of a project @Niraj-Kamdar, would need to be in a future release.

**Comment by cbrzn** [comment]
probably i am wrong but would it make sense to think that this can be achieved with polywrap re-imagine? @dOrgJelli @nerfZael

**Comment by dOrgJelli** [comment]
@cbrzn Yep! With rimagined it focuses on 2 things that related to this issue:
1. Stateful Wrapper Instances: wrappers can now be "stateful", instead of just being stateless functions. This means that you can instantiate a wrapper instance, call its methods, and it can retain some internal state within its heap memory.
2. Wrapper Instance References: wrapper instances can be passed as arguments to functions & wrapper instances. This allow you to effectively share the state of the wrapper instance (exposed through its methods) with the destination you're passing it to.

So, if we were to rework the above example in terms of reimagined using wrapper instances, it might look something like this (@nerfZael please confirm or deny):
```graphql
type LotsOfData @class {
  create(lotsOfBytes: Bytes!): Data! @staticFn

  length: UInt32! @fn

  readData(
    start: UInt32
    end: UInt32
  ): Bytes! @fn
}
```
and you can efficiently send around `LotsOfData` like so:
```graphql
type Module {
  someFunction(
    data: LotsOfData!
  ): boolean! @staticFn
}
```

NOTE: `someFunction` is receiving an instance reference as an argument, so we don't need to serialize all of the data when passing around the reference. It remains within the instance's memory.
